Hazel/src/Hazel/Debug/Instrumentor.h

```diff
@@ -1,129 +1,116 @@
+        std::ofstream m_OutputStream;
+        int m_ProfileCount;
+
+      public:
+        Instrumentor() : m_CurrentSession(nullptr), m_ProfileCount(0) {}
+
+        void BeginSession(const std::string &name, const std::string &filepath = "results.json") {
+            m_OutputStream.open(filepath);
+            WriteHeader();
+            m_CurrentSession = new InstrumentationSession{name};
+        }
+
+        void EndSession() {
+            WriteFooter();
+            m_OutputStream.close();
+            delete m_CurrentSession;
+            m_CurrentSession = nullptr;
+            m_ProfileCount = 0;
+        }
+
+        void WriteProfile(const ProfileResult &result) {
+            if (m_ProfileCount++ > 0) m_OutputStream << ",";
+
+            std::string name = result.Name;
+            std::replace(name.begin(), name.end(), '"', '\'');
+
+            m_OutputStream << "{";
+            m_OutputStream << "\"cat\":\"function\",";
+            m_OutputStream << "\"dur\":" << (result.End - result.Start) << ',';
+            m_OutputStream << "\"name\":\"" << name << "\",";
+            m_OutputStream << "\"ph\":\"X\",";
+            m_OutputStream << "\"pid\":0,";
+            m_OutputStream << "\"tid\":" << result.ThreadID << ",";
+            m_OutputStream << "\"ts\":" << result.Start;
+            m_OutputStream << "}";
+
+            m_OutputStream.flush();
+        }
+
+        void WriteHeader() {
+            m_OutputStream << "{\"otherData\": {},\"traceEvents\":[";
+            m_OutputStream.flush();
+        }
+
+        void WriteFooter() {
+            m_OutputStream << "]}";
+            m_OutputStream.flush();
+        }
+
+        static Instrumentor &Get() {
+            static Instrumentor instance;
+            return instance;
+        }
+    };
+
+    class InstrumentationTimer {
+      public:
+        InstrumentationTimer(const char *name) : m_Name(name), m_Stopped(false) {
+            m_StartTimepoint = std::chrono::high_resolution_clock::now();
+        }
+
+        ~InstrumentationTimer() {
+            if (!m_Stopped) Stop();
+        }
+
+        void Stop() {
+            auto endTimepoint = std::chrono::high_resolution_clock::now();
+
+            long long start =
+                std::chrono::time_point_cast<std::chrono::microseconds>(m_StartTimepoint).time_since_epoch().count();
+            long long end = std::chrono::time_point_cast<std::chrono::microseconds>(endTimepoint).time_since_epoch().count();
+
+            uint32_t threadID = std::hash<std::thread::id>{}(std::this_thread::get_id());
+            Instrumentor::Get().WriteProfile({m_Name, start, end, threadID});
+
+            m_Stopped = true;
+        }
+
+      private:
+        const char *m_Name;
+        std::chrono::time_point<std::chrono::high_resolution_clock> m_StartTimepoint;
+        bool m_Stopped;
+    };
+} // namespace Hazel
 
 #define HZ_PROFILE 1
 #if HZ_PROFILE
+	#ifndef __FUNCSIG__
```

Comment:
What about creating a macro `HZ_FUNC_SIG` (in core.h) as follows:

```cpp
#if defined(__FUNCSIG__)
	#define HZ_FUNC_SIG __FUNCSIG__
#elif defined (__PRETTY_FUNCTION__)
	#define HZ_FUNC_SIG __PRETTY_FUNCTION__
#else
	#error "No function signature function defined!"
#endif
```

_The macro name can be changed, the message here is about how the macro is defined (and where)._

Reply:
This might not work outside of a function as those macros are designed for usage inside a function, so besure to fully test it if going with this

Reply:
This is a nice and formalized method to solve this issue, however, it may be overkill to put it in core.h considering Cherno probably will only use HZ_FUNC_SIG inside of the instrumentor implementation. I like the idea of including a #error though and will update the code (as well as fix the whitespace)

Reply:
> Let me know what you think I should change

@GabeRundlett, now you define a HZ_FUNC_SIG when profile is set to 1, I would also add it in the else statement to be empty (like the other ones)? `	#define HZ_FUNC_SIG`
https://github.com/TheCherno/Hazel/blob/3aff8a00f063efb429d5e0e766820b9df9243207/Hazel/src/Hazel/Debug/Instrumentor.h#L127-L133

> This might not work outside of a function as those macros are designed for usage inside a function, so besure to fully test it if going with this

@ReDucTor, the place where this is used is literaly called `HZ_PROFILE_FUNCTION()`. I think it is up to the user to correctly utilize it? We have an alternative scope profiler as well in that case:
https://github.com/TheCherno/Hazel/blob/3aff8a00f063efb429d5e0e766820b9df9243207/Hazel/src/Hazel/Debug/Instrumentor.h#L136-L137

Reply:
It's not about where it's used but where you check those macros, which is outside a function.

Reply:
ohhhh, now I understand

Reply:
Resolved in https://github.com/TheCherno/Hazel/pull/174/commits/2b06109be382a435724c782d9f69ef8e228ec672